backend/app/treasury/adapters/covalent_pricefeed.py

```diff
@@ -0,0 +1,102 @@
+# pylint: disable=duplicate-code
+import datetime
+import json
+from os import getenv
+from typing import Any, TypeVar, Union
+
+import dateutil
+from celery.utils.log import get_logger
+from httpx import AsyncClient, HTTPStatusError, Timeout
+
+from ... import db
+from .types import Price
+
+COVALENT_URI = "https://api.covalenthq.com/v1"
+CACHE_HASH_PRICES = "covalent_prices"
+CACHE_KEY_TEMPLATE_PRICES = "{symbol}_{start}_{end}"
+
+RawPrices = TypeVar("RawPrices", list[dict[str, Any]], list[None])
+
+
+async def _get_pricing_data(
+    token_address: str,
+    start_date: datetime.datetime,
+    end_date: datetime.datetime,
+    chain_id: int = 1,
+) -> RawPrices:
+    start = start_date.strftime("%Y-%m-%d")
+    end = end_date.strftime("%Y-%m-%d")
+    timeout = Timeout(10.0, read=15.0, connect=30.0)
+    async with AsyncClient(
+        headers={
+            "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
+            + "AppleWebKit/537.36 (KHTML, like Gecko) "
+            + "Chrome/100.0.4896.127 Safari/537.36 Edg/100.0.1185.50"
+        },
+    ) as client:
+        if token_address == "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee":
+            token_address = "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"
+
+        req_url = (
+            COVALENT_URI
+            + f"/pricing/historical_by_addresses_v2/{chain_id}"
+            + f"/USD/{token_address}/?quote-currency=USD&format=JSON"
+            + f"&from={start}&to={end}&key=ckey_{getenv('COVALENT_KEY')}"
+        )
+
+        resp = await client.get(req_url, timeout=timeout)
+        resp.raise_for_status()
+        return resp.json()["data"][0]["prices"]
+
+
+async def get_token_hist_price_covalent(
+    token_address: str, token_symbol: str
+) -> list[Union[Price, None]]:
```

Comment:
I believe it's simply `list[Price]`.

Reply:
It can actually also return `[]` though `[None] != []` so I guess it should instead be `Union[list[Price], []]`.

Hopefully this provides better clarity.